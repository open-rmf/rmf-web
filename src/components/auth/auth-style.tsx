import { makeStyles } from '@material-ui/core';

const authStyles = makeStyles(theme => ({
  flexColumnContainer: {
    display: 'flex',
    flexDirection: 'column',
    alignItems: 'center',
    justifyContent: 'center',
  },
  fullPage: {
    width: '100%',
    height: '100%',
    backgroundColor: '#44497a',
  },
  authContainer: {
    flexDirection: 'column',
    borderStyle: 'none',
    borderRadius: 20,
    borderColor: 'black',
    padding: '70px',
    width: 'fit-content',
    minWidth: 250,
    backgroundColor: 'snow',
    boxShadow: theme.shadows[12],
    alignContent: 'center',
    justifyContent: 'center',
  },
  authTitle: {
    color: '#44497a',
  },
  authForm: {
    marginTop: 10,
    flexDirection: 'column',
  },
  buttonContainer: {
    border: '0',
    margin: '0.5rem',
    display: 'inline-flex',
    position: 'relative',
    minWidth: 0,
    flexDirection: 'column',
    verticalAlign: 'top',
    width: '100%',
    justifyContent: 'center',
    alignItems: 'center',
  },
  button: {
    width: '100%',
    backgroundColor: '#44497a',
  },
  logo: {
    width: 100,
    margin: '25px 0px 50px 0px',
  },
  resetPassword: {
    alignSelf: 'center',
    color: '#918e8e',
  },
}));

export default authStyles;
