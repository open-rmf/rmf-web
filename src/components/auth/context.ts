import React from 'react';
import Authenticator from './authenticator';

export const AuthContext = React.createContext<Authenticator | null>(null);

export default AuthContext;
