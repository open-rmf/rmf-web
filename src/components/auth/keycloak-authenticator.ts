import Keycloak_, { KeycloakInstance } from 'keycloak-js';
import Authenticator from './authenticator';

type ConfigType<T = typeof Keycloak_> = T extends (config: infer U) => unknown ? U : never;

export default class KeycloakAuthenticator implements Authenticator {
  static async create(config: ConfigType): Promise<KeycloakAuthenticator> {
    const inst = Keycloak_(config);
    const token = getLocalStorage('token');
    const idToken = getLocalStorage('idToken');
    const refreshToken = getLocalStorage('refreshToken');
    await inst.init({ token, idToken, refreshToken });
    try {
      await inst.updateToken(30);
    } catch {}
    console.log(inst.idToken);
    setOrClearLocalStorage('token', inst.token);
    setOrClearLocalStorage('idToken', inst.idToken);
    setOrClearLocalStorage('refreshToken', inst.refreshToken);
    return new KeycloakAuthenticator(inst);
  }

  get authenticated(): boolean {
    return !!this._inst.authenticated;
  }

  get sossToken(): string | undefined {
    return this._inst.idToken;
  }

  async login(redirectUri?: string): Promise<never> {
    await this._inst.login({ redirectUri });
    throw new Error('should not reach here');
  }

  async logout(): Promise<never> {
    localStorage.removeItem('token');
    localStorage.removeItem('idToken');
    localStorage.removeItem('refreshToken');
    this._inst.logout();
    throw new Error('should not reach here');
  }

  private constructor(private _inst: KeycloakInstance) {}
}

/**
 * Gets an item from localStorage, unlike `localStorage.getItem`, this returns undefined if item
 * does not exists.
 * @param key
 */
function getLocalStorage(key: string): string | undefined {
  const item = localStorage.getItem(key);
  return item ? item : undefined;
}

function setOrClearLocalStorage(key: string, value?: string): void {
  if (value === undefined) {
    localStorage.removeItem(key);
  } else {
    localStorage.setItem(key, value);
  }
}
