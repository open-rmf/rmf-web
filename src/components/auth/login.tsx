import { Button, Typography } from '@material-ui/core';
import React from 'react';
import { Redirect } from 'react-router';
import appConfig from '../../app-config';
import { UserContext } from '../../app-contexts';
import authStyles from './auth-style';

export default function Login(): React.ReactElement {
  const authenticator = appConfig.authenticator;
  const user = React.useContext(UserContext);
  const classes = authStyles();

  async function handleRmfLogin(_event: React.MouseEvent): Promise<void> {
    authenticator.login();
  }

  return user ? (
    <Redirect to="/" />
  ) : (
    <div className={`${classes.flexColumnContainer} ${classes.fullPage}`}>
      <div className={`${classes.flexColumnContainer} ${classes.authContainer}`}>
        <Typography variant="h4" className={classes.authTitle}>
          RoMi Dashboard
        </Typography>
        <img src="assets/ros-health.png" alt="" className={classes.logo} />
        <Button id="login-button" onClick={handleRmfLogin} variant="contained">
          Login with RMF
        </Button>
      </div>
    </div>
  );
}
