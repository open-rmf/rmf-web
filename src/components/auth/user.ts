export interface User {
  username: string;
  sossToken: string;
}
