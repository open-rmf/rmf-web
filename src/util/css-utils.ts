import * as L from 'leaflet';

export function viewBoxFromLeafletBounds(bounds: L.LatLngBoundsExpression): string {
  const lbounds = bounds instanceof L.LatLngBounds ? bounds : new L.LatLngBounds(bounds);
  const width = lbounds.getEast() - lbounds.getWest();
  const height = lbounds.getNorth() - lbounds.getSouth();
  return `0 0 ${width} ${height}`;
}

// colors used outside material ui
export const colorPalette: { [key: string]: string } = {
  unknown: '#cccccc',
};
