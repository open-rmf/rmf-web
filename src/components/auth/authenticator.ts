import debug from 'debug';
import EventEmitter from 'eventemitter3';
import Keycloak_, { KeycloakInstance } from 'keycloak-js';
import { User } from './user';

export type AuthenticatorEventType = {
  userChanged: [User | null];
};

export default interface Authenticator extends EventEmitter<AuthenticatorEventType> {
  readonly user?: User;
  readonly sossToken?: string;

  /**
   * It is required to call this before using any of the authenticator functions.
   */
  init(): Promise<void>;

  /**
   * Note: This redirects to external login page so it will never return.
   */
  login(): Promise<never>;

  /**
   * Note: This redirects to external logout page so it will never return.
   */
  logout(): Promise<void>;
}

type ConfigType<T = typeof Keycloak_> = T extends (config: infer U) => unknown ? U : never;

export class DefaultAuthenticator extends EventEmitter<AuthenticatorEventType>
  implements Authenticator {
  get user(): User | undefined {
    return this._user;
  }

  get sossToken(): string | undefined {
    return this._inst.idToken;
  }

  constructor(config: ConfigType, redirectUri?: string) {
    super();
    this._inst = Keycloak_(config);
    this._redirectUri = redirectUri;
  }

  async init() {
    debug.log('initializing authenticator');

    this._inst.onAuthSuccess = async () => {
      const profile = this._inst.profile || (await this._inst.loadUserProfile());
      this._user = {
        username: profile.username!,
      };
      debug.log('authenticated as', this._user.username);
      this.emit('userChanged', this._user);
    };

    this._inst.onAuthLogout = () => {
      debug.log('logout');
      this._user = undefined;
      this.emit('userChanged', null);
    };

    const token = getLocalStorage('token');
    const idToken = getLocalStorage('idToken');
    const refreshToken = getLocalStorage('refreshToken');
    await this._inst.init({ redirectUri: this._redirectUri, token, idToken, refreshToken });
    if (this._inst.authenticated) {
      await this._inst.loadUserProfile();
    }
    try {
      const refreshed = await this._inst.updateToken(30);
      refreshed && debug.log('token refreshed');
    } catch {}
    setOrClearLocalStorage('token', this._inst.token);
    setOrClearLocalStorage('idToken', this._inst.idToken);
    setOrClearLocalStorage('refreshToken', this._inst.refreshToken);
  }

  async login(): Promise<never> {
    await this._inst.login();
    throw new Error('should not reach here');
  }

  async logout(): Promise<never> {
    await this._inst.logout();
    throw new Error('should not reach here');
  }

  private _inst: KeycloakInstance;
  private _redirectUri?: string;
  private _user?: User;
}

/**
 * Gets an item from localStorage, unlike `localStorage.getItem`, this returns undefined if item
 * does not exists.
 * @param key
 */
function getLocalStorage(key: string): string | undefined {
  const item = localStorage.getItem(key);
  return item ? item : undefined;
}

function setOrClearLocalStorage(key: string, value?: string): void {
  if (value === undefined) {
    localStorage.removeItem(key);
  } else {
    localStorage.setItem(key, value);
  }
}
