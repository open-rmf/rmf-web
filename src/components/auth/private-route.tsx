import Debug from 'debug';
import React from 'react';
import { Redirect, Route, RouteProps, useLocation } from 'react-router';
import { LOGIN_ROUTE } from '../../util/url';
import { UserContext } from '../auth/contexts';
import Unauthorized from '../error-pages/unauthorized';

const debug = Debug('PrivateRoute');

interface Props extends React.PropsWithChildren<RouteProps> {
  // if true, do not redirect to login url if not authenticated
  noRedirectToLogin?: boolean;
}

/**
 * This component validates if the user is authenticated before rendering component passed as a
 * prop.
 */
const PrivateRoute = ({ noRedirectToLogin, children, ...rest }: Props): React.ReactElement => {
  const user = React.useContext(UserContext);
  const location = useLocation();

  function render(): React.ReactNode {
    if (user) {
      return children;
    } else {
      if (!noRedirectToLogin) {
        debug('accessing private route while unauthenticated');
        debug('redirecting to login page');
        return <Redirect to={{ pathname: LOGIN_ROUTE, state: { from: location } }} />;
      } else {
        return <Unauthorized />;
      }
    }
  }

  return <Route {...rest}>{render()}</Route>;
};

export default PrivateRoute;
