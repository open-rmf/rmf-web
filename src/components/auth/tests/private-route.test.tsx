import { createMount } from '@material-ui/core/test-utils';
import { createMemoryHistory, MemoryHistory } from 'history';
import React from 'react';
import { BrowserRouter, Router, Switch } from 'react-router-dom';
import { LOGIN_ROUTE } from '../../../util/url';
import { UserContext } from '../../auth/contexts';
import Unauthorized from '../../error-pages/unauthorized';
import PrivateRoute from '../private-route';

const mount = createMount();

describe('PrivateRoute', () => {
  let history: MemoryHistory;

  beforeEach(() => {
    history = createMemoryHistory();
    history.location.pathname = '/private';
  });

  test('renders correctly', () => {
    const component = mount(
      <BrowserRouter>
        <PrivateRoute path="/private" exact />
      </BrowserRouter>,
    );
    expect(component.html()).toMatchSnapshot();
  });

  test('redirects when unauthenticated', () => {
    mount(
      <Router history={history}>
        <Switch>
          <PrivateRoute path="/private" exact />
        </Switch>
      </Router>,
    );
    expect(history.location.pathname).toBe(LOGIN_ROUTE);
  });

  test('no redirects when authenticated', () => {
    mount(
      <UserContext.Provider value={{ username: 'test' }}>
        <Router history={history}>
          <Switch>
            <PrivateRoute path="/private" exact />
          </Switch>
        </Router>
      </UserContext.Provider>,
    );
    expect(history.location.pathname).toBe('/private');
  });

  test('shows unauthorized page when noRedirectToLogin is true', () => {
    const component = mount(
      <BrowserRouter>
        <PrivateRoute path="/private" exact noRedirectToLogin />
      </BrowserRouter>,
    );
    expect(history.location.pathname).toBe('/private');
    expect(component.find(Unauthorized)).toBeTruthy();
  });
});
