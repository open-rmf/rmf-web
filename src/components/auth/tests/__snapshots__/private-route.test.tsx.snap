// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PrivateRoute renders correctly 1`] = `""`;
